onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1688,6 +1689,133 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
                                   ONNX_NAMESPACE::matmulShapeInference(ctx, 0, 1);
                                 }));
 
+/**
+ * @brief Shape inference for MatMul with right hand side matrix quantized into int4
+ * @param ctx
+ * @param input_a_idx         points to the left hand size matrix input
+ * @param input_b_idx         points to the quantized right hand side matrix
+ * @param input_bshape_idx    points to the shape tensor of the right hand side matrix
+ */
+static void matmulQ4ShapeInference(ONNX_NAMESPACE::InferenceContext& ctx, int input_a_idx, int input_b_idx, int input_bshape_idx, MLAS_BLK_QUANT_TYPE blk_quant_type) {
+  if (!hasInputShape(ctx, input_a_idx) || !hasInputShape(ctx, input_b_idx)) {
+    return;
+  }
+
+  const auto& a_shape = ctx.getInputType(input_a_idx)->tensor_type().shape();
+  if (a_shape.dim_size() == 0) {
+    fail_shape_inference("Input tensors of wrong rank (0).");
+  }
+
+  const auto& blob_shape = ctx.getInputType(input_b_idx)->tensor_type().shape();
+  const auto& shape_shape = ctx.getInputType(input_bshape_idx)->tensor_type().shape();
+  if (shape_shape.dim_size() != 1 && shape_shape.dim(0).dim_value() != 2) {
```

Comment:
The logic is correct here, shape_shape should be a tensor that looks like: [2]

Reply:
I think this would also allow a shape of [3], for example, since `dim_size != 1` would be false.

Reply:
> I think this would also allow a shape of [3], for example, since `dim_size != 1` would be false.

how can it do that? the second half ensures that the value must be 2.

Reply:
I think you misunderstood the purpose of this variable here. This is describing the shape of the tensor that in turn describes the shape of the original B tensor shape. Since the packed B tensor lost its shape information, this operator has a parameter tensor that describes the shape of B.

Reply:
> how can it do that? the second half ensures that the value must be 2.

if the left side of the `&&` evaluates to false, the right side won't be evaluated.
https://en.wikipedia.org/wiki/Short-circuit_evaluation

my understanding is that we want to check that `shape_shape` has rank 1 (call this predicate X) and its only dimension has value 2 (call this Y). so we can do `if (!(X and Y)) { fail }` or equivalently, `if (!X or !Y) { fail }`.
https://en.wikipedia.org/wiki/De_Morgan%27s_laws